### DISCUSSION TITLE
Button Long Press  - Beta Test Group

### DISCUSSION THREAD

**Discussion by PRProd** [PRProd/HA-Firemote#477] [Announcements]
## 🎉 Beta Testing For Button Long Press Has Arrived! 🥳 
I have just published [the first beta version](https://github.com/PRProd/HA-Firemote/releases/tag/v4.0.0b1) of Firemote that supports button push and hold, also known as long press!  Now, I need your help to test it all out before the official release!
<br>

### How can you help?
After downloading the newest beta version, please compare the differences between how your physical remote reacts to long press events, and how Firemote reacts to those same events.  If everything is working properly, or you find an issue that's **not** mentioned in the "Known Issues / Limitations" section below, please let me know in this discussion thread right away.

I request that your reports contain the following information:
 - Media Player Device Type
 - Home Assistant version and Firemote beta version number
 - Dashboard interaction method
   - Web browser or mobile app
   - Browser and/or mobile app platform (Firefox on Windows, Chrome on Linux, iOS app, Android app, etc.)
 - Interaction details
   - What worked
   - What didn't work
     - Expected behavior vs Actual behavior
<br>

### Known issues / Limitations
This chart will be actively updated as testing continues.
| status id  | Fix Planned | Exists in                                                                        | Resolved in | Affected Devices  | Description                                     |
| --- | ----------- | -------------------------------------------------------------------------------- | ----------- | ----------------- | ----------------------------------------------- |
| ✅ ~~01~~  | ~~yes~~ | [HA 2024.8.0](https://www.home-assistant.io/blog/2024/08/07/release-20248/)      | [iPad App 2024.8](https://github.com/home-assistant/iOS/releases/tag/release%2F2024.8%2F2024.783) | ~~iPad OS Companion App~~ | ~~Firemote rendering is messed up / unusable - Follow the progress here https://github.com/PRProd/HA-Firemote/issues/478~~ |
| ▢ 02  | no          | [Firemote v4.0.0b1](https://github.com/PRProd/HA-Firemote/releases/tag/v4.0.0b1) | n/a         | All               | Built-in App launch buttons do not support long press |
| ▢ 03  | unknown     | [Firemote v4.0.0b1](https://github.com/PRProd/HA-Firemote/releases/tag/v4.0.0b1) |             | All               | button_overrides do not support long press |
| ▢ 04  | unknown     | [Firemote v4.0.0b1](https://github.com/PRProd/HA-Firemote/releases/tag/v4.0.0b1) |             | All               | custom_launchers do not support long press |
| ✅ ~~05~~  | ~~yes~~  | [Firemote v4.0.0b1](https://github.com/PRProd/HA-Firemote/releases/tag/v4.0.0b1) | [Firemote v4.0.0b3](https://github.com/PRProd/HA-Firemote/releases/tag/v4.0.0b3) | ~~All~~           | ~~DPad direction buttons (up, right, down, left) do not have long press support~~ |
| ▢ 06  | unknown     | [Firemote v4.0.0b1](https://github.com/PRProd/HA-Firemote/releases/tag/v4.0.0b1) |             | Roku              | Dpad center/select button hold is not supported.  Roku integration limitation - issue [123999](https://github.com/home-assistant/core/issues/123999) has been opened to address it |
| ✅ ~~07~~  |             | [Firemote v4.0.0b1](https://github.com/PRProd/HA-Firemote/releases/tag/v4.0.0b1) | [Firemote v4.0.0b3](https://github.com/PRProd/HA-Firemote/releases/tag/v4.0.0b3) | ~~Chromecast, Onn, Shield, Xiaomi~~ | ~~Dpad center/select button hold is not supported in "strong" mode unless an Android TV remote entity is defined~~ Compromised with a simulated repeat press of the button |
| ✅ ~~08~~  | ~~unknown~~ | [Firemote v4.0.0b1](https://github.com/PRProd/HA-Firemote/releases/tag/v4.0.0b1) | [Firemote v4.0.0b3](https://github.com/PRProd/HA-Firemote/releases/tag/v4.0.0b3) | ~~Roku~~          | ~~Replay button long press is not supported.  Roku integration limitation - issue [123999](https://github.com/home-assistant/core/issues/123999) has been opened to address it~~ Compromised with a simulated repeat press of the button |
| ✅ ~~09~~  | yes         | [Firemote v4.0.0b1](https://github.com/PRProd/HA-Firemote/releases/tag/v4.0.0b1) | [Firemote v4.0.0b4](https://github.com/PRProd/HA-Firemote/releases/tag/v4.0.0b4) | ~~All~~               | ~~Rewind & Fast Forward buttons do not have long press support~~ |
| ✅ ~~10~~  | yes         | [Firemote v4.0.0b1](https://github.com/PRProd/HA-Firemote/releases/tag/v4.0.0b1) | [Firemote v4.0.0b4](https://github.com/PRProd/HA-Firemote/releases/tag/v4.0.0b4) | ~~Most - Some Fire TV Devices gained support in beta 3~~ | ~~Volume up and Volume down buttons do not have long press support~~ |
| ✅ ~~11~~  | yes         | [Firemote v4.0.0b1](https://github.com/PRProd/HA-Firemote/releases/tag/v4.0.0b1) |  [Firemote v4.0.0b4](https://github.com/PRProd/HA-Firemote/releases/tag/v4.0.0b4) | ~~Amazon Fire, Roku~~ | ~~Channel up and Channel down buttons do not have long press support~~ |
| ▢ 12  | unknown     | [Firemote v4.0.0b1](https://github.com/PRProd/HA-Firemote/releases/tag/v4.0.0b1) |             | Amazon Fire       | programmable-one-button and programmable-two-button clicks haven't always worked properly for Amazon Fire style 6 remotes.  Long press event simulation for these buttons have been researched, but no resolution has been found |

<br>

Thank you all in advance for your help!  I know that many of you have been waiting a long time for this, so I appreciate your continued patience and partnership with this update.

<hr>

 ~ Tagging potentially interested users ~
@efiniste @mizuxe @electroflame @janstadt @dasadavid @darrenhomme
Sorry if I missed you and/or if I included you and you didn't want to be included

**Comment by qwerty13547** [comment]
Long press did not work for me. Let me know if there is something I need to change in the yaml if I use an existing Firemote remote setup. 

**Media Player Device Type:** Nvidia Sheild Pro (2019)
**Home Assistant version and Firemote beta version number:** 
Core: 2024.7.4
Supervisor: 2024.08.0
Operating System: 12.4
Frontend: 20240710.0
Firemote: v4.0.0.b1
**Dashboard interaction method:**
Webrowser Edge v127.0.2651.86 (64 bits) on win 10
Home Assistant App v2024.7.3 full on Android mobile running version 13.1.0.590(EX01) 
**Interaction details**
What didn't work: In Youtube app, press and hold (longpress) on forward and backward (D-pad) did not do anything more than before. One press and hold equals one press. 
Expected behavior: In Youtube app, press and hold (longpress) on forward and backward (D-pad) expected to fast forward / backward until button is released.

**Reply by PRProd** [reply]
@qwerty13547 ~  
Thank you for your report!

The DPad issue you reported was already in the chart above (id = 05).  This item as well as others noted with a "Fix Planned = yes" are expected to be finished before this new functionality is pushed into the official Firemote release.

**Reply by qwerty13547** [reply]
Sorry, I missed the big fat NOT in the sentence. I have to attribute that to a few days lack of sleep 🤦‍♂️

**Reply by PRProd** [reply]
No problem!  I REALLY appreciate that you took the time to try this beta!

**Reply by PRProd** [reply]
@qwerty13547 ~ [v4.0.0b3](https://github.com/PRProd/HA-Firemote/releases/tag/v4.0.0bb3) was just released, and it has that DPad support you were looking for.  😄

**Reply by qwerty13547** [reply]
Sweet! Good work👍

It work, but may still need some tweaking. 

Setup as above. 
When I compare long pressing the shield remote to Firemote, this is the outcome:

**Nvidia shield remote in YouTube app.:**
Long press scrolling sideways results in 5 videos per second. 
Continuing scrolling 1 video after releasing button. 
Scrolling engage almost immediately. 

Long press on center button on D pad open the action screen for the video.

**Firemote 4.0.03b in YouTube app:**
Long press scrolling sideways results in 1 video per second. 
Continuing scrolling 4-5 video after releasing button. 
Scrolling engage after 1 second.

Long press on center button on D pad does not work as intended, it starts playing video as a single press would. I imagine that it takes longer time to engage long press then Shield need to get the instruction to open the action screen for the video. So it se it as a single press.

**Reply by PRProd** [reply]
Thanks for the detail @qwerty13547!

>Firemote 4.0.03b in YouTube app:
>Long press scrolling sideways results in 1 video per second.
>Continuing scrolling 4-5 video after releasing button.
>Scrolling engage after 1 second.

Based on what you've experienced, am I correct to assume that you haven't connected to your Shield with the [Android TV Remote Integration](https://www.home-assistant.io/integrations/androidtv_remote/) as well?

If that's true, what's happening "under the hood" in this type of configuration is that Firemote is actually repeatedly sending the **ADB** command "left", "right", "up", or "down" 4 times every second.  ([actual code here](https://github.com/PRProd/HA-Firemote/blob/v4.0.0b3/dist/HA-Firemote.js#L5389-L5398))  No matter how fast or slow this command is repeated, it seems that the fastest response the Shield will provide is what you are seeing now.  It's slow, but that's why we have the [Android TV Remote Integration](https://www.home-assistant.io/integrations/androidtv_remote/)!  Adding this to your Firemote configuration will drastically improve performance!

help: [Why do I still need the ADB integration when there is an Android remote integration?](https://github.com/PRProd/HA-Firemote/wiki#why-do-i-still-need-the-adb-integration-when-there-is-an-android-remote-integration)

![image](https://github.com/user-attachments/assets/eb4be391-9518-4607-902a-c5e924b4f514)

When using both of them like this, the "long press" strategy changes.  In this configuration, we have a new option.  The [Android TV Remote Integration](https://www.home-assistant.io/integrations/androidtv_remote/) allows us to specify a duration of how long to hold the button down.  

After detecting a "long press" on the dashboard, Firemote [v4.0.0b3](https://github.com/PRProd/HA-Firemote/releases/tag/v4.0.0b3) sends the "up", "left", "right", or "down" command with a 0.5 second "hold down" time.  After 0.5 seconds has passed, Firemote checks to see if you are still holding the button down.  If so, it sends another 0.5 seconds hold down command to your Shield and will continue to check until you release the button. ([actual code here](https://github.com/PRProd/HA-Firemote/blob/v4.0.0b3/dist/HA-Firemote.js#L5379-L5388))

**Summary:**
If you do not add the [Android TV Remote Integration](https://www.home-assistant.io/integrations/androidtv_remote/), the speed you are seeing is the best it will ever get.  The BEST way to use Firemote with your Shield (with or without any "long press" detection) is to add the [Android TV Remote Integration](https://www.home-assistant.io/integrations/androidtv_remote/).
<br>

>Long press on center button on D pad does not work as intended, it starts playing video as a single press would. 

This is part of what was documented in the chart above at **id 07**.  When only an ADB entity is defined, there is no such thing as a "hold".  All of the documentation and [help threads](https://stackoverflow.com/a/27630157) say there should be, but for whatever reason it is ignored.  So, as noted above, I've opted to do the only thing I could: repeat the button press command 4 times per second.  This is why you saw your Shield behave as if the center button was just clicked. ([actual code here](https://github.com/PRProd/HA-Firemote/blob/v4.0.0b3/dist/HA-Firemote.js#L5686-L5697))

However, when adding the [Android TV Remote Integration](https://www.home-assistant.io/integrations/androidtv_remote/) we gain the ability to tell the Shield that we want a long press.  Pressing and holding the center button in Firemote with the Shield in my lab while using YouTube does work as expected. ([actual code here](https://github.com/PRProd/HA-Firemote/blob/v4.0.0b3/dist/HA-Firemote.js#L5682-L5685))
<br>

So @qwerty13547, I invite you to add the [Android TV Remote Integration](https://www.home-assistant.io/integrations/androidtv_remote/) to your Home Assistant, configure your Shield, and test once again!  THANKS AGAIN for your help with the testing here!

I hope to hear from you again soon!

-Doug

**Reply by PRProd** [reply]
@qwerty13547 ~ I'm curious if you've given this a try yet.  Do you have any updates?

**Comment by PRProd** [comment]
Firemote beta version [v4.0.0b2](https://github.com/PRProd/HA-Firemote/releases/tag/v4.0.0b2) was just released.

Nothing from the chart above has been updated since beta 1, but I would prefer that you test against this new beta 2 version.

**Comment by PRProd** [comment]
Firemote beta version [v4.0.0b3](https://github.com/PRProd/HA-Firemote/releases/tag/v4.0.0b3) has just been released.

This one has support for dpad longpress actions, as well as some important internal hints for your dashboard if you're experimenting with the new "sections" layouts.

Please see the updated chart at the top of this discussion for additional details.  I'd love to hear from you!

**Comment by Christarr2308** [comment]
Hi,
I have come across an issue with the firemote display on my iPad Pro 
However on my iPhone and windows PC it displays normally.
I have been using for around 3 months but the issue recently occurred in the last month.
I have tried to remove and reinstall the card but the problem still exists.
The scaling of the buttons are out.
B![image](https://github.com/user-attachments/assets/b402da0d-34f2-4a04-acb3-074d077d6ff8)
![image](https://github.com/user-attachments/assets/4801d846-4ebf-4d00-b29e-e82df0817bd7)

**Reply by PRProd** [reply]
Greetings @Christarr2308!

The problem you are describing was already captured in the chart above at **id 01**.  This is a known issue that was introduced with [HA 2024.08.0](https://www.home-assistant.io/blog/2024/08/07/release-20248/) and it mostly affects iPads.  The Firemote code isn't directly to blame for the change.

The good news is that for the past week, a group of very smart people have been working to track down the problem, and a fix for this issue should be published very soon.  

For more information, take a look here https://github.com/PRProd/HA-Firemote/issues/478

-Doug

**Reply by PRProd** [reply]
@Christarr2308 ~

As a followup, this rendering issue has been fixed in the latest iPad Companion app, [version 2024.8](https://github.com/home-assistant/iOS/releases/tag/release%2F2024.8%2F2024.783).  Simply download the [latest version from the App Store](https://apps.apple.com/us/app/home-assistant/id1099568401) if you haven't done so already, and you will be ready to go!

**Comment by PRProd** [comment]
## It's ready!!
Good news everyone!  Firemote beta version [v4.0.0b4](https://github.com/PRProd/HA-Firemote/releases/tag/v4.0.0b4) has just been released.  This latest beta contains all of the planned long press functionality, and completes the chart at the top!

This IS the release candidate!  Please download this new beta and let me know if you find anything wrong.  If not, it will be published soon!

Thanks in advance, everyone!

**Comment by PRProd** [comment]
## It's out!
I decided to launch it today.  😄 Long press support is now available in the official release of Firemote version [v4.0.0](https://github.com/PRProd/HA-Firemote/releases/tag/v4.0.0)
🎉

**Reply by yjamal01** [reply]
Doesn't seem to be working on firecube. A long press seems to be taken as multiple clicks of the button instead. Let me know if you need any other details and I will provide. Thanks again

**Reply by PRProd** [reply]
Thanks for the feedback @yjamal01!  Which button(s) are you referring to?

**Reply by yjamal01** [reply]
I first attmpted to long press the center button. This resulted in the TV registering multiple clicks of the center button (as if I was tapping it over and over in rapid succession) 
I also tried with the right arrow button, and got the same response.

**Reply by PRProd** [reply]
Gotcha!  I think what is happening is that you have your configuration set to Strong instead of the default for your device:
![image](https://github.com/user-attachments/assets/c91b6fab-e426-4c6a-9561-80b10b368a22)
instead of
![image](https://github.com/user-attachments/assets/bf0a2705-68a7-4119-bca2-49d4ac26d80c)

If that's the case, then the repeating button press is the only method that will work to simulate a "hold".  (ID 7 on the chart above)

The good news is that Amazon Fire products are the ONLY ones right now with actual support for "button down command" (wait for USER to release Firemote button "button up command" With [two exceptions](https://github.com/PRProd/HA-Firemote/wiki#amazon-fire), that family of devices actually has the BEST support for button hold!

Give "Default" a try, and let me know if that works better for you.

**Reply by yjamal01** [reply]
Ahh that makes sense, I do have strong set but its intentional.

On my firecube the default did not work. However following the instructions I found that "event 13" (iirc) worked. But then, as also mentioned in the documentation, when I would turn the tv on with the firemote, I would need to press a button on the physical remote before the dpad buttons would work.

I found that when using "strong" it would alleviate this limitation. Albeit its a little slower as also indicated, but definitely still usable. So for my usecase I would rather sacrifice the speed over the need to press the physical remote when turning on.

And even if the center press won't work on strong mode, I think for my usecase it still works better because then I don't have to try and find the remote =)

Although now that makes me think, what if I had a single button that sent a command in strong mode, while the rest of the buttons are "event 13"...I wonder if that would wake up the firemote or does it have to be a physical remote button press?

Either way thanks for your help, and please let me know if I misunderstood anything regarding the limitations and my preference

**Reply by PRProd** [reply]
You understand fully!  [Firemote is bound by the capabilities of the Home Assistant integrations](https://github.com/PRProd/HA-Firemote/wiki#how-does-firemote-send-commands-to-my-devices).

> Although now that makes me think, what if I had a single button that sent a command in strong mode, while the rest of the buttons are "event 13"...I wonder if that would wake up the firemote or does it have to be a physical remote button press?

I wish... [I tried and failed at that task](https://github.com/PRProd/HA-Firemote/discussions/115#discussioncomment-5970362)

**Reply by yjamal01** [reply]
Ahh damn, thanks for the info! I'll take not having to wake up the remote over having long press.
